File: 55_jump_game.py
def canJump(nums):
    from collections import deque

    # Function to check if the last index is reachable from the start index
    def can_reach_end(arr):
        max_reach = 0
        for i, num in enumerate(arr):
            if i > max_reach:
                return False
            max_reach = max(max_reach, i + num)
            if max_reach >= len(arr) - 1:
                return True
        return False

    if can_reach_end(nums):
        return True

    queue = deque([nums])  # Queue to store different possible arrays
    seen = set()  # To track visited arrays

    while queue:
        arr = queue.popleft()
        if tuple(arr) in seen:
            continue
        seen.add(tuple(arr))

        for i in range(1, arr[0] + 1):  # Generate new arrays
            new_arr = arr[i:]
            if new_arr and tuple(new_arr) not in seen:
                if can_reach_end(new_arr):
                    return True
                queue.append(new_arr)

    return False


# Taking user input
while True:
    try:
        nums = list(map(int, input("Enter a list of non-negative integers separated by spaces: ").split()))
        
        if any(n < 0 for n in nums):
            raise ValueError("All numbers must be non-negative integers.")
        
        if len(nums) == 0:
            raise ValueError("The list cannot be empty.")

        break  # Exit loop if input is valid

    except ValueError as e:
        print(f"Invalid input: {e}. Please try again.")

# Calling the function
print(canJump(nums))

File: 55_jump_game.js
function canJump(nums) {
  function canReachEnd(arr) {
    let maxReach = 0;
    for (let i = 0; i < arr.length; i++) {
      if (i > maxReach) return false;
      maxReach = Math.max(maxReach, i + arr[i]);
      if (maxReach >= arr.length - 1) return true;
    }
    return false;
  }

  if (canReachEnd(nums)) return true;

  let queue = [nums];
  let seen = new Set();

  while (queue.length > 0) {
    let arr = queue.shift();
    let key = arr.join(",");

    if (seen.has(key)) continue;
    seen.add(key);

    for (let i = 1; i <= arr[0]; i++) {
      let newArr = arr.slice(i);
      let newKey = newArr.join(",");

      if (newArr.length > 0 && !seen.has(newKey)) {
        if (canReachEnd(newArr)) return true;
        queue.push(newArr);
      }
    }
  }

  return false;
}

// Taking user input
const readline = require("readline");

const rl = readline.createInterface({
  input: process.stdin,
  output: process.stdout,
});

rl.question(
  "Enter a list of non-negative integers separated by spaces: ",
  (input) => {
    let nums = input.split(" ").map(Number);

    // Validate input
    if (nums.some((n) => isNaN(n) || n < 0)) {
      console.log("Invalid input: All numbers must be non-negative integers.");
    } else if (nums.length === 0) {
      console.log("Invalid input: The list cannot be empty.");
    } else {
      console.log(canJump(nums));
    }

    rl.close();
  }
);
